justfile

```
@@ -13,12 +13,41 @@ default_docker_tag := 'local'
 default_repo_name := 'ghcr.io/astriaorg'
 
 # Builds docker image for the crate. Defaults to 'local' tag.
-docker-build crate tag=default_docker_tag repo_name=default_repo_name:
-  docker buildx build --load --build-arg TARGETBINARY={{crate}} -f containerfiles/Dockerfile -t {{repo_name}}/$(echo {{crate}} | sed 's/^[^-]*-//'):{{tag}} .
-
-docker-build-and-load crate tag=default_docker_tag repo_name=default_repo_name:
-  @just docker-build {{crate}} {{tag}} {{repo_name}}
-  @just load-image $(echo {{crate}} | sed 's/^[^-]*-//') {{tag}} {{repo_name}}
+# NOTE: `_crate_short_name` is invoked as dependency of this command so that failure to pass a valid
+# crate will produce a maningful error message.
+docker-build crate tag=default_docker_tag repo_name=default_repo_name: (_crate_short_name crate "quiet")
+  #!/usr/bin/env sh
+  set -eu
+  short_name=`just _crate_short_name {{crate}}`
+  set -x
+  docker buildx build --load --build-arg TARGETBINARY={{crate}} -f containerfiles/Dockerfile -t {{repo_name}}/$short_name:{{tag}} .
+
+# Builds and loads docker image for the crate. Defaults to 'local' tag.
+# NOTE: `_crate_short_name` is invoked as dependency of this command so that failure to pass a valid
+# crate will produce a maningful error message.
+docker-build-and-load crate tag=default_docker_tag repo_name=default_repo_name: (_crate_short_name crate "quiet")
+  #!/usr/bin/env sh
+  set -eu
+  short_name=`just _crate_short_name {{crate}}`
+  set -x
+  just docker-build {{crate}} {{tag}} {{repo_name}}
+  just load-image $short_name {{tag}} {{repo_name}}
+
+# Maps a crate name to the shortened name used in the docker tag.
+# If `quiet` is an empty string the shortened name will be echoed. If `quiet` is a non-empty string,
+# the only output will be in the case of an error, where the input `crate` is not a valid one.
+_crate_short_name crate quiet="":
```

Instead of prefixing with `_`, maybe an attribute makes it clearer?
```suggestion
#[private]
crate_short_name crate:
```
You will note I also removed the `quiet` argument - I am not 100% we need it? I notice it is being used as an input to both `docker-*` recipes, but you also have `set -eu` to fail early. Can we just rely on that or is there extra use with having the arg?

---

We're using the leading underscore style elsewhere in this file rather than the `[private]` attribute, so I'll leave as is.

We need the quiet arg to silence the output on success.  The recipe is being invoked twice in each of the two parent recipes.

The first time it is being specified as a dependency so that it is invoked before running the parent.  This is purely to validate the specified crate is a supported one.  If that invocation fails, we get a clear error message and exit code 2.  In the success case, without the `quiet` arg, we'd get the shortened crate name printed to the console, which is confusing and needless noise.  So in that invocation, we set `quiet` to a non-empty string to suppress the output on success.

The second time it is called is in the body of the parent recipe, and in that case we require the successful output to populate our shell variable `short_name`, so we set `quiet` to an empty string.

Without the first call (when we call it as a dependency), if we only called in the body and the specified crate was invalid, we'd get no helpful error message.

---

Thank you for the explanation!